### PR TITLE
Save CLI arguments to the lock file

### LINKED ISSUE #308: Write `opam-monorepo` invocation into lock file
In #307 we discussed how to make the lockfiles reproducible in the presence of command line options (that are able to change the output of the lockfile in a non-trivial way). A compromise that would allow both people to overwrite values via the CLI and yet give hints how to reproduce the generated lockfile is to write the invocation of the `opam-monorepo` call that was used to create the lockfile into the lockfile itself.

The potential advantages are:

1. It allows people who want to reproduce the lock file to reproduce it. At least provided they use the same version of `opam-monorepo` but we don't really provide cross-version reproducibility for plenty of good reasons.
2.  It serves as documentation how the lockfile came to be, so newcomers to a project using `opam-monorepo` are aware of it and due to the invocation written there, they have a pointer to see how to use `opam-monorepo`

While this feature will not be implemented in #307, this issue was created to keep track of the idea.

### PR DESCRIPTION
**Pull request by NathanReb**
Fixes #308

### CODE REVIEW

File: lib/lockfile.ml [hunk 1]
```diff
@@ -261,6 +261,21 @@ module Depexts = struct
     List.concat all |> List.sort_uniq ~cmp:compare_elm
 end
 
+module Cli_args = struct
+  (** Field used to store the raw command line arguments passed to [lock].
+      It is set but not read and therefore is not part of the main lock file
+      type. *)
+
+  type _t = string list
```
Comment (Leonidas-from-XIV) [inline lib/lockfile.ml:269]: What is the reason to not name it `t`?

File: lib/lockfile.ml [hunk 2]
```diff
@@ -364,11 +379,12 @@ let from_opam ~opam_monorepo_cwd ?file opam =
       source_config;
     }
 
-let save ~opam_monorepo_cwd ~file t =
+let save ~opam_monorepo_cwd ~cli_args ~file t =
   let opam = to_opam ~opam_monorepo_cwd t in
+  let with_cli_args = Cli_args.add cli_args opam in
```
Comment (Leonidas-from-XIV) [inline lib/lockfile.ml:384]: Wouldn't it be nicer to do something like

```suggestion
  let opam = to_opam ~opam_monorepo_cwd t
    |> Cli_args.add cli_args
  in
```
Then it is clear that `opam` still contains a writeable opam file.

**Review by Leonidas-from-XIV — COMMENTED** [review]
Very nice. I have some nitpicks/clarifications but I think this is even better than what I envisioned in the original discussion.

File: lib/lockfile.ml (hunk 1 above)
Comment (NathanReb) [reply lib/lockfile.ml:269]: The compiler rightfully complains it is unused otherwise. I still wanted to keep the type to clarify things in the code.

**Review by Leonidas-from-XIV — APPROVED** [review]
:ship: it :) (After placating the formatter)

**Pull request merged**